app/src/main/java/com/asfoundation/wallet/transactions/PerkBonusAndGamificationService.kt

```diff
@@ -0,0 +1,298 @@
+        promotionsRepository.getGamificationStats(address),
+        promotionsRepository.getLevels(address),
+        getNewTransactions(address),
+        Function5 { lastShownLevel: Int, almostNextLevelLastShown: Int, stats: GamificationStats,
+                    allLevels: Levels, transactions: List<Transaction> ->
+          if (isCaseInvalidForNotifications(stats, transactions))
+            return@Function5
+          val maxBonus = allLevels.list.maxBy { level -> level.bonus }!!.bonus
+          val maxLevel = allLevels.list.maxBy { level -> level.level }!!.level
+          val currentLevel = stats.level
+          val bonusTransactionValue = getAllPerkBonusTransactionValues(transactions)
+          val currentLevelStartAmount = allLevels.list[currentLevel].amount
+          handlePerkAndLevelUpNotification(address, lastShownLevel, currentLevel, transactions,
+              stats, currentLevelStartAmount, maxBonus, maxLevel, bonusTransactionValue)
+          handleAlmostNextLevelNotification(address, almostNextLevelLastShown, currentLevel, stats,
+              currentLevelStartAmount, maxBonus, maxLevel)
+        }).doOnError { e -> e.printStackTrace() }.subscribeOn(Schedulers.io())
+  }
+
+  private fun handlePerkAndLevelUpNotification(address: String, lastShownLevel: Int,
+                                               currentLevel: Int, transactions: List<Transaction>,
+                                               stats: GamificationStats,
+                                               currentLevelStartAmount: BigDecimal,
+                                               maxBonus: Double, maxLevel: Int,
+                                               bonusTransactionValue: String) {
+    if (lastShownLevel < currentLevel && hasPurchaseResultedInLevelUp(transactions,
+            stats.totalSpend.minus(currentLevelStartAmount))) {
+      promotionsRepository.shownLevel(address, currentLevel,
+          GamificationContext.NOTIFICATIONS_LEVEL_UP)
+      buildNotification(createLevelUpNotification(stats, maxBonus,
+          currentLevel == maxLevel, bonusTransactionValue),
+          NOTIFICATION_SERVICE_ID_PERK_AND_LEVEL_UP)
+    } else if (bonusTransactionValue.isNotEmpty()) {
+      buildNotification(createPerkBonusNotification(bonusTransactionValue),
+          NOTIFICATION_SERVICE_ID_PERK_AND_LEVEL_UP)
+    }
+  }
+
+  private fun handleAlmostNextLevelNotification(address: String, almostNextLevelLastShown: Int,
+                                                currentLevel: Int, stats: GamificationStats,
+                                                currentLevelStartAmount: BigDecimal,
+                                                maxBonus: Double, maxLevel: Int) {
+    if (currentLevel in (almostNextLevelLastShown + 1) until maxLevel) {
+      // the current level being shown is the value to be stored regarding if the notification
+      // for almost reaching next level (current level + 1) has been shown or not
```

Comment:
typo I think: Regardless?
Tip: Normally when there's the need to comment on what something does, it means that we can improve it (readability by extracting, etc). This can be done here imo, by extracting the if condition to a private method and/or by extracting what's inside the if to a private method. Exception to the rule when the comment it's to alert someone of possible problems, why it was done that way or public API documentation